File: serial/embedded/SerialEmbedded/FletcherChecksum.cpp

#include "FletcherChecksum.h";

/**
 * https://en.wikipedia.org/wiki/Fletcher%27s_checksum
 * 
 * 8-bit implementation (16-bit checksum)
 * 
 * This implementation was not written with optimization
 * in mind. There are ways to optimize this code if needed
*/

namespace FletcherChecksum
{
uint16_t generate(const uint8_t *src, const size_t sz)
{
  uint16_t sum0 = 0;
  uint16_t sum1 = 0;

  for (size_t i = 0; i < sz; i++)
  {
    sum0 = (sum0 + src[i]) % 255;
    sum1 = (sum1 + sum0) % 255;
  }

  return (sum1 << 8) | sum0;
}

void append(uint8_t *src, size_t sz, uint16_t sum)
{
  const uint8_t sum0 = sum & 255;
  const uint8_t sum1 = (sum >> 8) & 255;

  const uint8_t chk0 = 255 - ((sum0 + sum1) % 255);
  const uint8_t chk1 = 255 - ((sum0 + chk0) % 255);

  const size_t fst = sz - 2;
  const size_t lst = sz - 1;

  src[fst] = chk0;
  src[lst] = chk1;
}

boolean valid(const uint8_t *src, const size_t sz)
{
  return generate(src, sz) == 0;
}

//======================================
// Tests
//======================================

boolean test_generate_2()
{
  // Arrange
  const uint8_t msg[] = {0x01, 0x02};

  // Act
  const uint16_t sum = generate(msg, 2);

  // Assert
  const boolean pass = sum == 0x0403;

  return pass;
}

boolean test_generate_5()
{
  // Arrange
  const uint8_t msg[] = "abcde";

  // Act
  const uint16_t sum = generate(msg, 5);

  // Assert
  const boolean pass = sum == 0xC8F0;

  return pass;
}

boolean test_generate_6()
{
  // Arrange
  const uint8_t msg[] = "abcdef";

  // Act
  const uint16_t sum = generate(msg, 6);

  // Assert
  const boolean pass = sum == 0x2057;

  return pass;
}

boolean test_generate_8()
{
  // Arrange
  const uint8_t msg[] = "abcdefgh";

  // Act
  const uint16_t sum = generate(msg, 8);

  // Assert
  const boolean pass = sum == 0x0627;

  return pass;
}

boolean test_append_2()
{
  // Arrange
  const uint16_t sum = 0x0403;

  uint8_t msg[4];
  msg[0] = 0x01;
  msg[1] = 0x02;

  // Act
  append(msg, 4, sum);

  // Assert
  return msg[2] == 0xF8 && msg[3] == 0x04;
}

boolean test_valid_2()
{
  // Arrange
  uint8_t msg[] = {0x01, 0x02, 0xF8, 0x04};

  // Act
  const boolean isValid = valid(msg, 4);

  // Assert
  return isValid;
}
};

File: serial/plugin/droplit-plugin-openag-serial/FletcherChecksum.js
"use strict";
/**
 * https://en.wikipedia.org/wiki/Fletcher%27s_checksum
 *
 * 8-bit implementation (16-bit checksum)
 *
 * This implementation was not written with optimization
 * in mind. There are ways to optimize this code if needed
*/
exports.__esModule = true;
var FletcherChecksum;
(function (FletcherChecksum) {
    function generate(src) {
        var sum0 = 0;
        var sum1 = 0;
        src.forEach(function (elem) {
            sum0 = (sum0 + elem) % 255;
            sum1 = (sum1 + sum0) % 255;
        });
        return new Buffer([sum1, sum0]);
    }
    FletcherChecksum.generate = generate;
    function append(src, sum) {
        var sum0 = sum[1];
        var sum1 = sum[0];
        var chk0 = 255 - ((sum0 + sum1) % 255);
        var chk1 = 255 - ((sum0 + chk0) % 255);
        var len = src.length;
        var res = new Buffer(len + 2);
        for (var i = 0; i < len; i++) {
            res[i] = src[i];
        }
        res[len] = chk0;
        res[len + 1] = chk1;
        return res;
    }
    FletcherChecksum.append = append;
    function strip(src) {
        return Buffer.from(src.slice(0, src.length - 2));
    }
    FletcherChecksum.strip = strip;
    function valid(src) {
        var sum = generate(src);
        return sum[0] == 0 && sum[1] == 0;
    }
    FletcherChecksum.valid = valid;
    //======================================
    // Tests
    //====================================== 
    function test_generate_2() {
        // Arrange
        var msg = new Buffer([0x01, 0x02]);
        // Act
        var sum = generate(msg);
        // Assert
        var pass = sum[0] == 0x04 && sum[1] == 0x03;
        return pass;
    }
    FletcherChecksum.test_generate_2 = test_generate_2;
    function test_generate_5() {
        // Arrange ("abcde")
        var msg = new Buffer([97, 98, 99, 100, 101]);
        // Act
        var sum = generate(msg);
        // Assert
        var pass = sum[0] == 0xC8 && sum[1] == 0xF0;
        return pass;
    }
    FletcherChecksum.test_generate_5 = test_generate_5;
    function test_generate_6() {
        // Arrange ("abcdef")
        var msg = new Buffer([97, 98, 99, 100, 101, 102]);
        // Act
        var sum = generate(msg);
        // Assert
        var pass = sum[0] == 0x20 && sum[1] == 0x57;
        return pass;
    }
    FletcherChecksum.test_generate_6 = test_generate_6;
    function test_generate_8() {
        // Arrange ("abcdefgh")
        var msg = new Buffer([97, 98, 99, 100, 101, 102, 103, 104]);
        // Act
        var sum = generate(msg);
        // Assert
        var pass = sum[0] == 0x06 && sum[1] == 0x27;
        return pass;
    }
    FletcherChecksum.test_generate_8 = test_generate_8;
    function test_append_2() {
        // Arrange ("abcdefgh")
        var sum = new Buffer([0x04, 0x03]);
        var msg = new Buffer([0x01, 0x02]);
        // Act
        var data = append(msg, sum);
        // Assert
        var pass = data[2] == 0xF8 && data[3] == 0x04;
        return pass;
    }
    FletcherChecksum.test_append_2 = test_append_2;
    function test_strip_2() {
        // Arrange
        var data = new Buffer([0x01, 0x02, 0xF8, 0x04]);
        // Act
        var msg = strip(data);
        // Assert
        return msg.length == 2 && msg[0] == 0x01 && msg[1] == 0x02;
    }
    FletcherChecksum.test_strip_2 = test_strip_2;
    function test_valid_2() {
        // Arrange
        var msg = new Buffer([0x01, 0x02, 0xF8, 0x04]);
        // Act
        var isValid = valid(msg);
        // Assert
        return isValid;
    }
    FletcherChecksum.test_valid_2 = test_valid_2;
})(FletcherChecksum = exports.FletcherChecksum || (exports.FletcherChecksum = {}));
